service/src/main/java/io/camunda/service/DocumentServices.java

```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH under
+ * one or more contributor license agreements. See the NOTICE file distributed
+ * with this work for additional information regarding copyright ownership.
+ * Licensed under the Camunda License 1.0. You may not use this file
+ * except in compliance with the Camunda License 1.0.
+ */
+package io.camunda.service;
+
+import io.camunda.search.clients.CamundaSearchClient;
+import io.camunda.service.security.auth.Authentication;
+import io.camunda.service.transformers.ServiceTransformers;
+import io.camunda.zeebe.broker.client.api.BrokerClient;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+
+public class DocumentServices extends ApiServices<DocumentServices> {
+
+  // TODO: This is an in-memory implementation, replace it with a real document store
+  private static final String STORE_ID = "default";
+  private final Map<UUID, byte[]> documents = new HashMap<>();
+
+  public DocumentServices(final BrokerClient brokerClient, final CamundaSearchClient searchClient) {
+    this(brokerClient, searchClient, null, null);
+  }
+
+  public DocumentServices(
+      final BrokerClient brokerClient,
+      final CamundaSearchClient searchClient,
+      final ServiceTransformers transformers,
+      final Authentication authentication) {
+    super(brokerClient, searchClient, transformers, authentication);
+  }
+
+  @Override
+  public DocumentServices withAuthentication(final Authentication authentication) {
+    return new DocumentServices(brokerClient, searchClient, transformers, authentication);
+  }
+
+
+  public CompletableFuture<DocumentReferenceResponse> createDocument(
+      DocumentCreateRequest request, InputStream contentInputStream) {
+
+    validateStoreId(request.storeId);
+
+    final String id = request.documentId != null ? request.documentId : UUID.randomUUID().toString();
+    final byte[] content;
```

Comment:
What about returning an `Either<DocumentCreationError, DocumentCreationResult>` to provide more information in case of an error?

Reply:
Not sure, actually. All services we've implemented so far either return the success type or fail with an exception. Callers can handle this accordingly. An `Either` would not be false per se but makes the service API too complex. A service method is fine to throw an exception since it shouldn't be the "user-facing" method, it's used by something like a REST gateway that decides how to handle the exception.

All data needed by consumers can be provided in the exception, I would say. If there is more than a message to convey, create a specific exception with additional attributes that can be formatted by the consumer accordingly. Just my 2 cents, though.

Reply:
Ok, I thought we might want to make use of a nice and typesafe error hierarchy using `sealed` classes or interfaces. The nice thing is that this would allow clients (the web-layer) to map each error to the right response in a fully typesafe way. Introducing a new error requires the client to handle it and map it properly.

Reply:
I see what you mean 👍 I am not completely opposed to that. Would be nice to handle this in a follow-up issue (to keep the focus here). We can explore that issue and create a PR with a proposal on how it could look. I'd like to have multiple eyes on such a decision from different teams since it affects multiple teams in how they write their services API.

Reply:
I used this pattern in the past (using kotlin sealed classes) and it was a joy to use. As suggested lets follow-up in a separate conversation.